### ISSUE TITLE
async_accept returning EAGAIN/EWOULDBLOCK

### ISSUE THREAD

**Issue by acmorrow** [chriskohlhoff/asio#237]
I just tried upgrading ASIO from 524288cb4fcf84664b3dc39cb4424c7509969b92 to 862aed305dcf91387535519c9549c17630339a12. Surprisingly, that change seems to have broken async_accept.

Given a basic_socket_acceptor, and one on which non_blocking(true) has been called, we are still finding that calls to basic_socket_acceptor::async_accept is returning with EAGAIN/EWOULDBLOCK. Since our application re-issues the call to async_accept in the handler for async_accept, this results in the application spinning.

Reverting ASIO back to 524288cb4fcf84664b3dc39cb4424c7509969b92 restores the expected behavior of async_accept.

I can try to create a smaller repro than building all of mongodb, but first I'd like to check whether this is a known issue, or whether a quick inspection of that range of commits reveals anything suspect.

To be clear, the *only* change in the tree is the upgrade of ASIO.

**Comment by acmorrow** [comment]
I have reproduced this. The following C++14 program accepts connections on 524288c, but crashes in in the abort inside accept_handler on 862aed3:

```
#include <asio.hpp>

#include <thread>

void do_accept(asio::basic_socket_acceptor<asio::generic::stream_protocol>& acceptor) {

    auto shared_sock = std::make_shared<asio::generic::stream_protocol::socket>(
        acceptor.get_io_context());

    auto& created_sock = *shared_sock;
    auto accept_handler = [&acceptor, shared_sock = std::move(shared_sock)](std::error_code ec) {
        if (ec) abort();
        std::printf("Accepted connection\n");
        do_accept(acceptor);
        shared_sock->close();
    };

    acceptor.async_accept(created_sock, std::move(accept_handler));
}

int main(int argc, char* argv[]) {

    asio::io_context context;

    asio::generic::stream_protocol::endpoint endpoint =
        asio::ip::tcp::endpoint(asio::ip::address::from_string("0.0.0.0"), 12345);

    asio::basic_socket_acceptor<asio::generic::stream_protocol> acceptor(context);
    acceptor.open(endpoint.protocol());
    acceptor.set_option(decltype(acceptor)::reuse_address(true));

    std::error_code ec;
    acceptor.non_blocking(true, ec);
    if (ec) abort();

    acceptor.bind(endpoint, ec);
    if (ec) abort();

    std::thread driver([&context] {
        const auto work = asio::make_work_guard(context);
        context.run();
    });

    acceptor.listen();
    do_accept(acceptor);
    driver.join();

    return 0;
}

```

**Comment by acmorrow** [comment]
Note that commenting out the call to `non_blocking` puts this back to working correctly, but it definitely should be legal to set the acceptor non-blocking and then call async_accept on it.

**Comment by chriskohlhoff** [comment]
Fixed in 230c0d2ae035c5ce1292233fcab03cea0d341264. This was a latent bug that has been there for a very long time, but was suppressed in your particular use case by the bug that was fixed in 48e7ef864f66f6e8403be2fb4c1f63d21a5902e6.

**Comment by acmorrow** [comment]
Thanks for the quick fix. I'm running it through our regression suite now.
